src/features/listVisualizer/tree/Tree.tsx

```diff
@@ -0,0 +1,459 @@
+          ? constructTree(headNode)
+          : isFunction(headNode)
+          ? constructFunction(headNode)
+          : constructData(headNode);
+      }
+
+      if (visitedStructures.indexOf(tailNode) > -1) {
+        // tree already built
+        node.right = visitedStructures.indexOf(tailNode);
+      } else {
+        node.right = isPair(tailNode)
+          ? constructTree(tailNode)
+          : isFunction(tailNode)
+          ? constructFunction(tailNode)
+          : constructData(tailNode);
+      }
+
+      return node;
+    }
+
+    /**
+     * Returns a node representing the given function.
+     * Also memoizes the function object, for the case where the
+     * function appears multiple times in the data structure.
+     * @param {Function} func The function to construct a node for.
+     */
+    function constructFunction(func: Function) {
+      const node = new FunctionTreeNode(nodeCount);
+
+      // memoise current function
+      visitedStructures[nodeCount] = func;
+      treeNodes[nodeCount] = node;
+      nodeCount++;
+
+      return node;
+    }
+
+    /**
+     * Returns a node representing the given data.
+     * Anything except functions and pairs are considered data, including empty lists.
+     * @param {Data} data The data to construct a node for.
+     */
+    function constructData(data: Data) {
+      return new DataTreeNode(data);
+    }
+
+    const visitedStructures: (Function | Pair)[] = []; // detects cycles
+    const treeNodes: DrawableTreeNode[] = [];
+    const rootNode = constructTree(tree);
+
+    return new Tree(rootNode, treeNodes);
+  }
+
+  draw(x: number, y: number): JSX.Element {
+    return new TreeDrawer(this).draw(x, y);
+  }
+}
+
+/**
+ *  Drawer function of a tree
+ */
+class TreeDrawer {
+  private tree: Tree;
+
+  // keeps track the extreme left end of the tree. In units of pixels.
+  private minLeft = 500;
+
+  private drawables: JSX.Element[];
+
+  constructor(tree: Tree) {
+    this.tree = tree;
+    this.drawables = [];
+  }
+
+  /**
+   *  Draws a tree at x, y, by calling drawNode on the root at x, y.
+   */
+  draw(x: number, y: number): JSX.Element {
+    this.drawNode(this.tree.rootNode, x, y, x, y);
+    return (
+      <Layer offsetX={this.minLeft - 20} offsetY={0}>
+        {this.drawables}
+      </Layer>
+    );
+  }
+
+  /**
+   *  Draws the box for the pair representing the tree, then recursively draws its children.
+   *  A slash is drawn for empty lists.
+   *
+   *  If a child node has been drawn previously, an arrow is drawn pointing to the children,
+   *  instead of drawing the child node again.
+   * @param {TreeNode} node The node to draw.
+   * @param {number} x The x position to draw at.
+   * @param {number} y The y position to draw at.
+   * @param {number} parentX The x position of the parent. If there is no parent, it is the same as x.
+   * @param {number} parentY The y position of the parent. If there is no parent, it is the same as y.
```

Comment:
The `{TreeNode}` and `{number}` in the function documentation doesn't really do anything